### PR TITLE
R4R: Add timeouts to http servers

### PR DESCRIPTION
**Pull request by zmanian**
The current http server implementation does not enforce anytime outs on connections. This facilitates file descriptor exhaustion.

This introduces timeouts into the http server implementation.

### CODE REVIEW

File: rpc/lib/server/http_server.go [hunk 1]
```diff
@@ -55,13 +55,14 @@ func StartHTTPServer(
...
-
 	go func() {
-		err := http.Serve(
-			listener,
-			RecoverAndLogHandler(maxBytesHandler{h: handler, n: maxBodyBytes}, logger),
-		)
-		logger.Info("RPC HTTP server stopped", "err", err)
+		s := &http.Server{
+			Handler:        RecoverAndLogHandler(maxBytesHandler{h: handler, n: maxBodyBytes}, logger),
+			ReadTimeout:    10 * time.Second,
```
Comment (zmanian) [inline rpc/lib/server/http_server.go]: Is 10 seconds too long?

**Comment by melekes** [comment]
```
==================
WARNING: DATA RACE
Write at 0x00c420316270 by goroutine 31:
  github.com/tendermint/tendermint/rpc/lib/server.StartHTTPAndTLSServer.func1()
      /go/src/github.com/tendermint/tendermint/rpc/lib/server/http_server.go:112 +0x259

Previous write at 0x00c420316270 by goroutine 32:
  github.com/tendermint/tendermint/rpc/lib/server.StartHTTPAndTLSServer()
      /go/src/github.com/tendermint/tendermint/rpc/lib/server/http_server.go:119 +0xd7f
  github.com/tendermint/tendermint/rpc/lib/server.TestStartHTTPAndTLSServer()
      /go/src/github.com/tendermint/tendermint/rpc/lib/server/http_server_test.go:74 +0xe0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 31 (running) created at:
  github.com/tendermint/tendermint/rpc/lib/server.StartHTTPAndTLSServer()
      /go/src/github.com/tendermint/tendermint/rpc/lib/server/http_server.go:105 +0xaa8
  github.com/tendermint/tendermint/rpc/lib/server.TestStartHTTPAndTLSServer()
      /go/src/github.com/tendermint/tendermint/rpc/lib/server/http_server_test.go:74 +0xe0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 32 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:112 +0x324
==================
--- FAIL: TestStartHTTPAndTLSServer (0.00s)
	Error Trace:	http_server_test.go:75
	Error:      	Expected nil, but got: &netutil.limitListener{Listener:(*net.TCPListener)(0xc4202c0040), sem:(chan struct {})(0xc4202c44e0), closeOnce:sync.Once{m:sync.Mutex{state:0, sema:0x0}, done:0x0}, done:(chan struct {})(0xc4202c4540)}
	Test:       	TestStartHTTPAndTLSServer
	testing.go:730: race detected during execution of test
FAIL
```

File: rpc/lib/server/http_server.go (hunk 1 above)
Comment (melekes) [reply rpc/lib/server/http_server.go]: I'd lower both to 3s

```suggestion
			ReadTimeout:    3 * time.Second,
```

File: rpc/lib/server/http_server.go [hunk 2]
```diff
@@ -55,13 +55,14 @@ func StartHTTPServer(
...
+			WriteTimeout:   10 * time.Second,
+			MaxHeaderBytes: 1 << 20,
```
Comment (melekes) [inline rpc/lib/server/http_server.go]: default. can be removed

```suggestion
```

File: rpc/lib/server/http_server.go [hunk 3]
```diff
@@ -101,13 +102,16 @@ func StartHTTPAndTLSServer(
 	if config.MaxOpenConnections > 0 {
 		listener = netutil.LimitListener(listener, config.MaxOpenConnections)
 	}
+	go func() {
+		s := &http.Server{
+			Handler:        RecoverAndLogHandler(maxBytesHandler{h: handler, n: maxBodyBytes}, logger),
+			ReadTimeout:    10 * time.Second,
+			WriteTimeout:   10 * time.Second,
+			MaxHeaderBytes: 1 << 20,
```
Comment (melekes) [inline rpc/lib/server/http_server.go]: default. can be removed

```suggestion
```

File: rpc/lib/server/http_server.go [hunk 4]
```diff
@@ -55,13 +55,14 @@ func StartHTTPServer(
...
 	go func() {
-		err := http.Serve(
-			listener,
-			RecoverAndLogHandler(maxBytesHandler{h: handler, n: maxBodyBytes}, logger),
-		)
-		logger.Info("RPC HTTP server stopped", "err", err)
+		s := &http.Server{
+			Handler:        RecoverAndLogHandler(maxBytesHandler{h: handler, n: maxBodyBytes}, logger),
+			ReadTimeout:    10 * time.Second,
+			WriteTimeout:   10 * time.Second,
```
Comment (melekes) [inline rpc/lib/server/http_server.go]:
```suggestion
			WriteTimeout:   3 * time.Second,
```

**Comment by jaekwon** [comment]
This has magic constants in several places, and I'm not sure how it relates to other timeout parameters that we already have for peer connections.  Needs a bit of review to parameterize & make consistent w/ the rest of the codebase.

**Comment by ebuchman** [comment]
> not sure how it relates to other timeout parameters that we already have for peer connections. 

What does this have to do with peer connections? Are you saying we should make the http and p2p connection timeouts somehow related?

> See main thread

What is this referring to ?

**Comment by jaekwon** [comment]
> What does this have to do with peer connections? Are you saying we should make the http and p2p connection timeouts somehow related?

Ah no, we shouldn't do that.  Lets just use constants instead of 10, and otherwise good to go.

**Comment by ebuchman** [comment]
Oh - this breaks `/broadcast_tx_commit`. If committing the tx takes longer than the timeout here, then once the commit finally happens we get `curl: (52) Empty reply from server`.

Not sure how to reconcile ... 

Does anyone know if/what the issue was here in the first place? Is there any endpoint that can hang for a while that we'd need this for?

**Comment by melekes** [comment]
> Oh - this breaks /broadcast_tx_commit. If committing the tx takes longer than the timeout here, then once the commit finally happens we get curl: (52) Empty reply from server.

`/broadcast_tx_commit` timeout is 10 sec. now https://github.com/tendermint/tendermint/blob/d8ab8509de625c222f23d4f54ec817f3ddfcb83e/rpc/core/mempool.go#L197. We can make it a bit smaller or increase read/write timeouts.

File: rpc/lib/server/http_server.go [hunk 5]
```diff
@@ -27,16 +27,25 @@ const (
 	// maxBodyBytes controls the maximum number of bytes the
 	// server will read parsing the request body.
 	maxBodyBytes = int64(1000000) // 1MB
+
+	// same as the net/http default
+	maxHeaderBytes = 1 << 20
```
Comment (melekes) [inline rpc/lib/server/http_server.go]: Go will use a default in the absence of such. Is there a reason to duplicate on our side?

Comment (ebuchman) [reply rpc/lib/server/http_server.go]: Just to be explicit.

**Comment by ebuchman** [comment]
> We can make it smaller or increase read/write timeouts (to 10 sec.)

We need to make one a function of the other so this doesn't become some kind of implicit dependency between constants.

I also still want to understand why we need these timeouts in the first place - under what circumstance are they actually important to us ?

**Comment by ebuchman** [comment]
Ok, I read https://blog.cloudflare.com/exposing-go-on-the-internet/ and https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/

It's clear we need to do this to handle hostile clients, but I still wonder if the original issue that prompted this might lie deeper than this fix, assuming our clients were functioning properly. I suspect there are other issues with file descriptors in Tendermint, but anyways.

@zmanian @melekes PTAL.

**Comment by zmanian** [comment]
I think the other file descriptor issues might a result of leveldb settings

**Pull request merged**